### PR TITLE
More checks of sample-sheet

### PR DESCRIPTION
**Pull request by asp8200**
Two checks for ensuring that the pipeline stops with a meaningful error message if

1. the sample-sheet only contains normal-samples, but some of the requested tools require tumor-samples, and
2. the sample-sheet only contains tumor-samples, but some of the requested tools require normal-samples.

I did some test-runs of the this with sample-sheet containing

1. only normal-samples,
2. only tumor-samples, and
3. a mix of normal samples and tumor-samples.

@lassefolkersen might also do some test-runs for these new checks.

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
    - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/sarek/tree/master/.github/CONTRIBUTING.md)
    - [ ] If necessary, also make a PR on the nf-core/sarek _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint .`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: workflows/sarek.nf [hunk 1]
```diff
@@ -1034,9 +1034,14 @@ def extract_csv(csv_file) {
             }
         }
 
+    numberOfSamples = 0
+    numberOfNormalSamples = 0
+    numberOfTumorSamples = 0
```
Comment (maxulysse) [inline workflows/sarek.nf:1039]:
```suggestion
    sample_count_all = 0
    sample_count_normal = 0
    sample_count_tumor = 0
```

File: workflows/sarek.nf [hunk 2]
```diff
@@ -1066,6 +1072,34 @@ def extract_csv(csv_file) {
...
+            def requested_tools_requiring_tumor_samples = []
+            tools_requiring_tumor_samples.each{ tool_requiring_tumor_samples ->
+                if (params.tools.contains(tool_requiring_tumor_samples)) requested_tools_requiring_tumor_samples.add(tool_requiring_tumor_samples)
+            }
+            if (!requested_tools_requiring_tumor_samples.isEmpty()) {
+                log.error('The sample-sheet only contains normal-samples, but the following tools, which were requested by the option "tools", expect at least one tumor-sample : ' + requested_tools_requiring_tumor_samples.join(", "))
+                System.exit(1)
+            }
+        } else if ((numberOfTumorSamples == numberOfSamples) && params.tools) {  // In this case, the sample-sheet contains no normal/germline-samples
+            def tools_requiring_normal_samples = ['ascat', 'deepvariant', 'haplotypecaller']
```
[the remark this answers is not in the document]
Comment (asp8200) [reply workflows/sarek.nf:1092]: @maxulysse : Are you sure about this? Judging from the diagram below, `ascat` requires both normal-samples and tumor-samples:

image

**Comment by maxulysse** [comment]
Can you update the CHANGELOG?

File: workflows/sarek.nf [hunk 3]
```diff
@@ -1066,6 +1072,34 @@ def extract_csv(csv_file) {
...
+        if ((sample_count_normal == sample_count_all) && params.tools) { // In this case, the sample-sheet contains no tumor-samples
+            def tools_requiring_tumor_samples = ['ascat', 'controlfreec', 'mutect2', 'msisensorpro']
+            def requested_tools_requiring_tumor_samples = []
+            tools_requiring_tumor_samples.each{ tool_requiring_tumor_samples ->
+                if (params.tools.contains(tool_requiring_tumor_samples)) requested_tools_requiring_tumor_samples.add(tool_requiring_tumor_samples)
+            }
+            if (!requested_tools_requiring_tumor_samples.isEmpty()) {
+                log.error('The sample-sheet only contains normal-samples, but the following tools, which were requested by the option "tools", expect at least one tumor-sample : ' + requested_tools_requiring_tumor_samples.join(", "))
+                System.exit(1)
+            }
```
Comment (maxulysse) [inline workflows/sarek.nf:1090]:
```suggestion
            def tools_tumor = ['ascat', 'controlfreec', 'mutect2', 'msisensorpro']
            def tools_tumor_asked = []
            tools_tumor.each{ tool ->
                if (params.tools.contains(tool)) tools_tumor_asked.add(tool)
            }
            if (!tools_tumor_asked.isEmpty()) {
                log.error('The sample-sheet only contains normal-samples, but the following tools, which were requested by the option "tools", expect at least one tumor-sample : ' + tools_tumor_asked.join(", "))
                System.exit(1)
            }
```

Comment (maxulysse) [reply workflows/sarek.nf:1090]: might be easier to read with simpler variable names?
What do you think?

File: workflows/sarek.nf [hunk 4]
```diff
@@ -1066,6 +1072,34 @@ def extract_csv(csv_file) {
...
+            def tools_requiring_normal_samples = ['ascat', 'deepvariant', 'haplotypecaller']
+            def requested_tools_requiring_normal_samples = []
+            tools_requiring_normal_samples.each{ tool_requiring_normal_samples ->
+                if (params.tools.contains(tool_requiring_normal_samples)) requested_tools_requiring_normal_samples.add(tool_requiring_normal_samples)
+            }
+            if (!requested_tools_requiring_normal_samples.isEmpty()) {
+                log.error('The sample-sheet only contains tumor-samples, but the following tools, which were requested by the option "tools", expect at least one normal-sample : ' + requested_tools_requiring_normal_samples.join(", "))
+                System.exit(1)
+            }
+        }
```
Comment (maxulysse) [inline workflows/sarek.nf:1101]:
```suggestion
        } else if ((sample_count_tumor == sample_count_all) && params.tools) {  // In this case, the sample-sheet contains no normal/germline-samples
            def tools_normal = ['ascat', 'deepvariant', 'haplotypecaller']
            def tools_normal_asked = []
            tools_normal.each{ tool ->
                if (params.tools.contains(tool)) tools_normal_asked.add(tool)
            }
            if (!tools_normal_asked.isEmpty()) {
                log.error('The sample-sheet only contains tumor-samples, but the following tools, which were requested by the option "tools", expect at least one normal-sample : ' + tools_normal_asked.join(", "))
                System.exit(1)
            }
        }
```

**Pull request merged**